From 40e75dafb479269aeb8d426bd2db31e7522d3287 Mon Sep 17 00:00:00 2001
From: Peter Varga <pvarga@inf.u-szeged.hu>
Date: Fri, 30 May 2025 15:34:34 +0200
Subject: [PATCH 10/13] UPSTREAM: NativeSkiaOutputDevice: Set color type
 according to NativePixmap support

NativePixmapEGLX11Binding requires the BGRA_8888 color type, use it for
the check too. The GL graphics context doesn't necessarily mean that we
need BGRA.

Pick-to: 6.10
Change-Id: I41a4f1f6da94c10f8e6843e6edfed913bb98fabd
Reviewed-by: Moss Heim <moss.heim@qt.io>
---
 .../compositor/native_skia_output_device.cpp  | 15 +++++++++++
 .../native_skia_output_device_opengl.cpp      |  5 ++--
 .../src/core/ozone/gl_ozone_angle_qt.cpp      | 27 ++++++-------------
 .../src/core/ozone/gl_ozone_angle_qt.h        | 20 ++++++++++++++
 4 files changed, 46 insertions(+), 21 deletions(-)

diff --git a/qtwebengine/src/core/compositor/native_skia_output_device.cpp b/qtwebengine/src/core/compositor/native_skia_output_device.cpp
index f42b2b9334..948686be7a 100644
--- a/qtwebengine/src/core/compositor/native_skia_output_device.cpp
+++ b/qtwebengine/src/core/compositor/native_skia_output_device.cpp
@@ -20,6 +20,8 @@
 #include "ui/gl/gl_fence.h"
 
 #if BUILDFLAG(IS_OZONE)
+#include "ozone/gl_ozone_angle_qt.h"
+
 #include "ui/ozone/public/ozone_platform.h"
 #endif
 
@@ -63,6 +65,19 @@ NativeSkiaOutputDevice::NativeSkiaOutputDevice(
                                         .supports_native_pixmaps;
     qCDebug(lcWebEngineCompositor, "Native Buffer Supported: %s",
             m_isNativeBufferSupported ? "yes" : "no");
+
+    auto typeToString = [](ui::NativePixmapSupportType type) -> const char * {
+        switch (type) {
+        case ui::NativePixmapSupportType::kDMABuf:
+            return "DMABuf";
+        case ui::NativePixmapSupportType::kX11Pixmap:
+            return "X11Pixmap";
+        default:
+            return "None";
+        }
+    };
+    qCDebug(lcWebEngineCompositor, "Native Pixmap Support Type: %s",
+            typeToString(ui::GLOzoneANGLEQt::getNativePixmapSupportType()));
 #endif
 }
 
diff --git a/qtwebengine/src/core/compositor/native_skia_output_device_opengl.cpp b/qtwebengine/src/core/compositor/native_skia_output_device_opengl.cpp
index 7cfc390d8a..46eb337b44 100644
--- a/qtwebengine/src/core/compositor/native_skia_output_device_opengl.cpp
+++ b/qtwebengine/src/core/compositor/native_skia_output_device_opengl.cpp
@@ -12,6 +12,7 @@
 
 #if BUILDFLAG(IS_OZONE)
 #include "ozone/gl_helper.h"
+#include "ozone/gl_ozone_angle_qt.h"
 #include "ozone/ozone_util_qt.h"
 
 #include "base/posix/eintr_wrapper.h"
@@ -62,8 +63,8 @@ NativeSkiaOutputDeviceOpenGL::NativeSkiaOutputDeviceOpenGL(
     qCDebug(lcWebEngineCompositor, "Native Skia Output Device: OpenGL");
 
     SkColorType skColorType = kRGBA_8888_SkColorType;
-#if BUILDFLAG(IS_OZONE_X11) && QT_CONFIG(xcb_glx_plugin)
-    if (OzoneUtilQt::usingGLX() && m_contextState->gr_context_type() == gpu::GrContextType::kGL)
+#if BUILDFLAG(IS_OZONE)
+    if (ui::GLOzoneANGLEQt::getNativePixmapSupportType() == ui::NativePixmapSupportType::kX11Pixmap)
         skColorType = kBGRA_8888_SkColorType;
 #endif
 
diff --git a/qtwebengine/src/core/ozone/gl_ozone_angle_qt.cpp b/qtwebengine/src/core/ozone/gl_ozone_angle_qt.cpp
index 359d04a454..b5fa45cd89 100644
--- a/qtwebengine/src/core/ozone/gl_ozone_angle_qt.cpp
+++ b/qtwebengine/src/core/ozone/gl_ozone_angle_qt.cpp
@@ -6,6 +6,7 @@
 // found in the LICENSE file.
 
 #include "gl_ozone_angle_qt.h"
+#include "surface_factory_qt.h"
 
 #include "ui/base/ozone_buildflags.h"
 #include "ui/gl/gl_bindings.h"
@@ -26,23 +27,12 @@ extern __eglMustCastToProperFunctionPointerType EGL_GetProcAddress(const char *p
 }
 
 namespace ui {
-namespace {
-// Based on //ui/ozone/platform/x11/x11_surface_factory.cc
-enum class NativePixmapSupportType {
-    // Importing native pixmaps not supported.
-    kNone,
-
-    // Native pixmaps are imported directly into EGL using the
-    // EGL_EXT_image_dma_buf_import extension.
-    kDMABuf,
-
-    // Native pixmaps are first imported as X11 pixmaps using DRI3 and then into
-    // EGL.
-    kX11Pixmap,
-};
-
-NativePixmapSupportType GetNativePixmapSupportType()
+
+NativePixmapSupportType GLOzoneANGLEQt::getNativePixmapSupportType()
 {
+    if (!QtWebEngineCore::SurfaceFactoryQt::SupportsNativePixmaps())
+        return NativePixmapSupportType::kNone;
+
     if (gl::GLSurfaceEGL::GetGLDisplayEGL()->ext->b_EGL_EXT_image_dma_buf_import)
         return NativePixmapSupportType::kDMABuf;
 
@@ -53,7 +43,6 @@ NativePixmapSupportType GetNativePixmapSupportType()
 
     return NativePixmapSupportType::kNone;
 }
-} // namespace
 
 bool GLOzoneANGLEQt::LoadGLES2Bindings(const gl::GLImplementationParts & /*implementation*/)
 {
@@ -107,7 +96,7 @@ gl::EGLDisplayPlatform GLOzoneANGLEQt::GetNativeDisplay()
 
 bool GLOzoneANGLEQt::CanImportNativePixmap(gfx::BufferFormat format)
 {
-    switch (GetNativePixmapSupportType()) {
+    switch (getNativePixmapSupportType()) {
     case NativePixmapSupportType::kDMABuf:
         return NativePixmapEGLBinding::IsBufferFormatSupported(format);
 #if BUILDFLAG(IS_OZONE_X11)
@@ -125,7 +114,7 @@ GLOzoneANGLEQt::ImportNativePixmap(scoped_refptr<gfx::NativePixmap> pixmap,
                                    gfx::Size plane_size, const gfx::ColorSpace &color_space,
                                    GLenum target, GLuint texture_id)
 {
-    switch (GetNativePixmapSupportType()) {
+    switch (getNativePixmapSupportType()) {
     case NativePixmapSupportType::kDMABuf:
         return NativePixmapEGLBinding::Create(pixmap, plane_format, plane, plane_size, color_space,
                                               target, texture_id);
diff --git a/qtwebengine/src/core/ozone/gl_ozone_angle_qt.h b/qtwebengine/src/core/ozone/gl_ozone_angle_qt.h
index e8d7af6aa7..240fa98b6c 100644
--- a/qtwebengine/src/core/ozone/gl_ozone_angle_qt.h
+++ b/qtwebengine/src/core/ozone/gl_ozone_angle_qt.h
@@ -1,6 +1,10 @@
 // Copyright (C) 2024 The Qt Company Ltd.
 // SPDX-License-Identifier: LicenseRef-Qt-Commercial OR LGPL-3.0-only OR GPL-2.0-only OR GPL-3.0-only
 
+// Copyright 2016 The Chromium Authors
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
 #ifndef GL_OZONE_ANGLE_QT_H
 #define GL_OZONE_ANGLE_QT_H
 
@@ -8,9 +12,25 @@
 
 namespace ui {
 
+// Based on //ui/ozone/platform/x11/x11_surface_factory.cc
+enum class NativePixmapSupportType {
+    // Importing native pixmaps not supported.
+    kNone,
+
+    // Native pixmaps are imported directly into EGL using the
+    // EGL_EXT_image_dma_buf_import extension.
+    kDMABuf,
+
+    // Native pixmaps are first imported as X11 pixmaps using DRI3 and then into
+    // EGL.
+    kX11Pixmap,
+};
+
 class GLOzoneANGLEQt : public GLOzoneEGL
 {
 public:
+    static NativePixmapSupportType getNativePixmapSupportType();
+
     bool InitializeStaticGLBindings(const gl::GLImplementationParts &implementation) override;
     bool InitializeExtensionSettingsOneOffPlatform(gl::GLDisplay *display) override;
     scoped_refptr<gl::GLSurface> CreateViewGLSurface(gl::GLDisplay *display,
-- 
2.51.0

